Nits on "parallel" flag help message and documentation

sickpig opened the pull request.

sickpig committed:
Change `parallel` help message according to the fact it's flag not a parameter

ptschip commented on src/parallel.cpp:
```diff
@@ -20,7 +20,8 @@
 
-static const unsigned int nScriptCheckQueues = 4;
+static const unsigned int nScriptCheckQueues = std::min(GetNumCores(), 4);
+
```
actually this needs to be 4...even though you may only have 2 cores we still need to be able to handle 4 concurrent blocks.  So it will be a bit slower on smaller cored systems but most serious miners will have 4 or more cores these days.

ptschip replied:
there should probably be a comment here about it.  You can maybe refer to the high level design doc  bu-parallel-validation.md in the /docs folder.

sickpig replied:
looking at the doc now. Do you have tl;dr explanation why we need 4 cores rather than, e.g., 6?

ptschip commented:
3 probably would be ok, but I thought to be conservative to use 4.  no
other reason really, and i think 5 would be way overkill.



On 26/06/2018 7:54 AM, Andrea Suisani wrote:
>
> *@sickpig* commented on this pull request.
>
> ------------------------------------------------------------------------
>
> In src/parallel.cpp
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1173#discussion_r198177538>:
>
> > @@ -20,7 +20,8 @@
>  
>  using namespace std;
>  
> -static const unsigned int nScriptCheckQueues = 4;
> +static const unsigned int nScriptCheckQueues = std::min(GetNumCores(), 4);
> +
>
> looking at the doc now. Do you have tl;dr explanation why we need 4
> cores rather than, e.g., 6?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1173#discussion_r198177538>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0My8fCIFjsj20PfEQ554w6v6TJPNks5uAksTgaJpZM4U37Fm>.
>

sickpig commented:
@ptschip 
what about something like: 

```c++
std::max(std::max(4, GetNumCores() - x))
```

where `x` is the number of cores to spare to avoid an overkill.

I looked at the code and I realized that `MAX_SCRIPTCHECK_THREADS` lost is meaning since we are going to cap the max number of validation check to `nScriptCheckQueues` (4) due to: 

```c++
CParallelValidation::CParallelValidation(int threadCount, boost::thread_group *threadGroup)
    : semThreadCount(nScriptCheckQueues)
{
    // A single thread has no parallelism so just use the main thread.  Equivalent to parallel being turned off.
    if (threadCount <= 1)
        threadCount = 0;
    else if (threadCount > MAX_SCRIPTCHECK_THREADS)
        threadCount = MAX_SCRIPTCHECK_THREADS;
    nThreads = threadCount;

    LOGA("Using %d threads for script verification\n", threadCount);

    while (QueueCount() < nScriptCheckQueues)
    {
        auto queue = new CCheckQueue<CScriptCheck>(128);

        for (unsigned int i = 0; i < nThreads; i++)
            threadGroup->create_thread(boost::bind(&AddScriptCheckThreads, i + 1, queue));

        vQueues.push_back(queue);
    }
}
```

ptschip commented:
script check threads are different from scriptcheck queues,...each
script check queue has it's own set of script check threads.  

not sure how std::max(std::max(4, GetNumCores() - x)) will help...won't
getnumcores() -x always be zero? at any rate in the worse case if there
are more queues than cores then I'm not sure I see any serious issue
with that. There will be some additional overhead with more thread
switching going on but at least all 4 blocks will continue to
process...there may be what a 5 or 10% extra overhead?

another point although minor is that the parallel.py tests have a 4
block slow to validate race in there which will likely not work at times
if we only have a smaller number of queues...that could be fixed not
doubt but will also be a bit of work.

On 26/06/2018 12:15 PM, Andrea Suisani wrote:
>
> @ptschip <https://github.com/ptschip>
> what about something like:
>
> std::max(std::max(4, GetNumCores() - x))
>
> where |x| is the number of cores to spare to avoid an overkill.
>
> I looked at the code and I realized that |MAX_SCRIPTCHECK_THREADS|
> lost is meaning since we are going to cap the max number of validation
> check to |nScriptCheckQueues| (4) due to:
>
> CParallelValidation::CParallelValidation(int threadCount, boost::thread_group *threadGroup)
>     : semThreadCount(nScriptCheckQueues)
> {
>     // A single thread has no parallelism so just use the main thread.
> Equivalent to parallel being turned off.
>     if (threadCount <= 1)
>         threadCount = 0;
>     else if (threadCount > MAX_SCRIPTCHECK_THREADS)
>         threadCount = MAX_SCRIPTCHECK_THREADS;
>     nThreads = threadCount;
>
>     LOGA("Using %d threads for script verification\n", threadCount);
>
>     while (QueueCount() < nScriptCheckQueues)
>     {
>         auto queue = new CCheckQueue<CScriptCheck>(128);
>
>         for (unsigned int i = 0; i < nThreads; i++)
>             threadGroup->create_thread(boost::bind(&AddScriptCheckThreads, i + 1, queue));
>
>         vQueues.push_back(queue);
>     }
> }
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1173#issuecomment-400430616>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AMRF0Bmsih2D4QBj_BK8gD-ks6pLrcx7ks5uAohWgaJpZM4U37Fm>.
>

sickpig commented:
@ptschip 

> script check threads are different from scriptcheck queues,...each script check queue has it's own set of script check threads.

fair enough, thanks for the explanation. 


`(std::max(4, GetNumCores() - x)` would be equal to 4 till `GetNumCores() - x` is < 4, once it gets over 4 then we are going to spin up as much threads as `GetNumCores() -x`, which is not 0. 

> won't getnumcores() -x always be zero

Suppose you have a machine with 10 cores and you chose `x` as 50% of the total number of cores. That mean that you would use 5 threads for PV.

Wonder that a better way to set `nScriptCheckQueues` would be something like: 

```c++
`(std::max(4, GetNumCores() * x)`
```

where `x` belongs to `[0.5,1]`.

sickpig commented:
going to substitute the second commit w/ one that just add a comment to explain why we chose 4 as the default and not tweakable number of PV queues.

sickpig committed:
Add a reference to doc/bu-parallel-validation.md above nScriptCheckQueues init

sickpig committed:
Formatting of doc/bu-parallel-validation.md

Mainly removing trailing spaces and add missing back trick around
function name. Also clarified what we meant "wrong chain' in the last
paragraph adding "chain with lowest amount of accumulated POW"

[skip ci]

sickpig commented:
Update: 

remove the change to `nScriptCheckQueues` init value. Added two doc commits:

- Formatting of doc/bu-parallel-validation.md
- Add a reference to doc/bu-parallel-validation.md above `nScriptCheckQueues` initialization

ptschip commented on doc/bu-parallel-validation.md:
```diff
@@ -85,29 +85,29 @@ mine themself.
 
 b) The following attack is outlined, however after some discussion, it was found not to be possible.  The reason being that
-PV will only be happening on the chainactive tip which means that the blocks being validated will all have the same parent.
+PV will only be happening on the `chainaAtive` tip which means that the blocks being validated will all have the same parent.
```
-chainaAtive
+chainActive

sickpig commented:
ooops sorry. you are right. fixing

On Thu, Jun 28, 2018 at 2:00 PM, Peter Tschipper <notifications@github.com>
wrote:

> *@ptschip* commented on this pull request.
> ------------------------------
>
> In doc/bu-parallel-validation.md
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1173#discussion_r198814362>
> :
>
> >
>  b) The following attack is outlined, however after some discussion, it was found not to be possible.  The reason being that
> -PV will only be happening on the chainactive tip which means that the blocks being validated will all have the same parent.
> +PV will only be happening on the `chainaAtive` tip which means that the blocks being validated will all have the same parent.
>
> -chainaAtive
> +chainActive
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/BitcoinUnlimited/BitcoinUnlimited/pull/1173#pullrequestreview-132809845>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/ABZrE5bTA3jsAIjadjQ1EU89jmMbbqRFks5uBMVUgaJpZM4U37Fm>
> .
>
